File: homework-1/backend-dotnet/src/BankingApi/Services.cs
namespace BankingApi;

public class TransactionService
{
    private readonly TransactionStore _store;

    public TransactionService(TransactionStore store) => _store = store;

    public Transaction Create(CreateTransactionRequest r)
    {
        var transaction = new Transaction(
            Guid.NewGuid().ToString(),
            r.FromAccount!,
            r.ToAccount!,
            r.Amount!.Value,
            r.Currency!,
            Enum.Parse<TransactionType>(r.Type!, ignoreCase: true),
            DateTimeOffset.UtcNow,
            TransactionStatus.Completed);
        return _store.Save(transaction);
    }

    public IReadOnlyList<Transaction> FindAll() => _store.FindAll();

    public Transaction? FindById(string id) => _store.FindById(id);

    public IReadOnlyList<Transaction> Find(string? accountId, string? type, DateOnly? from, DateOnly? to)
    {
        return _store.FindAll().Where(t =>
            (accountId is null || t.FromAccount == accountId || t.ToAccount == accountId) &&
            (type is null || t.Type.ToString().ToLowerInvariant() == type) &&
            (from is null || DateOnly.FromDateTime(t.Timestamp.UtcDateTime) >= from) &&
            (to is null || DateOnly.FromDateTime(t.Timestamp.UtcDateTime) <= to)
        ).ToList();
    }
}

public class AccountService
{
    private readonly TransactionStore _store;

    public AccountService(TransactionStore store) => _store = store;

    public decimal BalanceOf(string account)
    {
        decimal balance = 0m;
        foreach (var t in _store.FindAll())
        {
            if (t.Status != TransactionStatus.Completed) continue;
            switch (t.Type)
            {
                case TransactionType.Deposit:
                    if (t.ToAccount == account) balance += t.Amount;
                    break;
                case TransactionType.Withdrawal:
                    if (t.FromAccount == account) balance -= t.Amount;
                    break;
                case TransactionType.Transfer:
                    if (t.FromAccount == account) balance -= t.Amount;
                    if (t.ToAccount == account) balance += t.Amount;
                    break;
            }
        }
        return balance;
    }

    public SummaryResponse SummaryOf(string account)
    {
        decimal deposits = 0m, withdrawals = 0m;
        int count = 0;
        DateTimeOffset? mostRecent = null;

        foreach (var t in _store.FindAll())
        {
            if (t.FromAccount != account && t.ToAccount != account) continue;
            count++;
            if (mostRecent is null || t.Timestamp > mostRecent) mostRecent = t.Timestamp;
            if (t.Status == TransactionStatus.Completed)
            {
                if (t.Type == TransactionType.Deposit && t.ToAccount == account) deposits += t.Amount;
                else if (t.Type == TransactionType.Withdrawal && t.FromAccount == account) withdrawals += t.Amount;
            }
        }
        return new SummaryResponse(deposits, withdrawals, count, mostRecent);
    }

    public InterestResponse Interest(string account, decimal rate, int days)
    {
        if (rate < 0) throw new ArgumentException("rate must be zero or positive");
        if (days <= 0) throw new ArgumentException("days must be a positive integer");

        var balance = BalanceOf(account);
        var interest = Math.Round(balance * rate * days / 365m, 2, MidpointRounding.AwayFromZero);
        return new InterestResponse(account, balance, rate, days, interest);
    }
}

File: homework-1/backend-fastapi/app/services.py
import uuid
from datetime import date, datetime, timezone
from decimal import ROUND_HALF_UP, Decimal
from typing import Optional

from .models import (
    InterestResponse,
    SummaryResponse,
    Transaction,
    TransactionRequest,
    TransactionStatus,
    TransactionType,
)
from .store import TransactionStore

_ZERO = Decimal("0")
_CENTS = Decimal("0.01")
_YEAR = Decimal("365")


class TransactionService:
    def __init__(self, store: TransactionStore) -> None:
        self.store = store

    def create(self, req: TransactionRequest) -> Transaction:
        transaction = Transaction(
            id=str(uuid.uuid4()),
            fromAccount=req.fromAccount,
            toAccount=req.toAccount,
            amount=req.amount,
            currency=req.currency,
            type=req.type,
            timestamp=datetime.now(timezone.utc),
            status=TransactionStatus.completed,
        )
        return self.store.save(transaction)

    def find_all(self) -> list[Transaction]:
        return self.store.find_all()

    def find_by_id(self, transaction_id: str) -> Optional[Transaction]:
        return self.store.find_by_id(transaction_id)

    def find(
        self,
        account_id: Optional[str] = None,
        tx_type: Optional[str] = None,
        date_from: Optional[date] = None,
        date_to: Optional[date] = None,
    ) -> list[Transaction]:
        result: list[Transaction] = []
        for t in self.store.find_all():
            if account_id and t.fromAccount != account_id and t.toAccount != account_id:
                continue
            if tx_type and t.type.value != tx_type:
                continue
            day = t.timestamp.date()
            if date_from and day < date_from:
                continue
            if date_to and day > date_to:  # inclusive upper bound
                continue
            result.append(t)
        return result


class AccountService:
    def __init__(self, store: TransactionStore) -> None:
        self.store = store

    def balance_of(self, account_id: str) -> Decimal:
        balance = _ZERO
        for t in self.store.find_all():
            if t.status != TransactionStatus.completed:
                continue
            if t.type == TransactionType.deposit and t.toAccount == account_id:
                balance += t.amount
            elif t.type == TransactionType.withdrawal and t.fromAccount == account_id:
                balance -= t.amount
            elif t.type == TransactionType.transfer:
                if t.fromAccount == account_id:
                    balance -= t.amount
                if t.toAccount == account_id:
                    balance += t.amount
        return balance

    def summary_of(self, account_id: str) -> SummaryResponse:
        deposits = _ZERO
        withdrawals = _ZERO
        count = 0
        most_recent: Optional[datetime] = None
        for t in self.store.find_all():
            if t.fromAccount != account_id and t.toAccount != account_id:
                continue
            count += 1
            if most_recent is None or t.timestamp > most_recent:
                most_recent = t.timestamp
            if t.status == TransactionStatus.completed:
                if t.type == TransactionType.deposit and t.toAccount == account_id:
                    deposits += t.amount
                elif t.type == TransactionType.withdrawal and t.fromAccount == account_id:
                    withdrawals += t.amount
        return SummaryResponse(
            totalDeposits=deposits,
            totalWithdrawals=withdrawals,
            transactionCount=count,
            mostRecentTransactionDate=most_recent,
        )

    def interest(self, account_id: str, rate: Decimal, days: int) -> InterestResponse:
        if rate < 0:
            raise ValueError("rate must be zero or positive")
        if days <= 0:
            raise ValueError("days must be a positive integer")
        balance = self.balance_of(account_id)
        value = (balance * rate * Decimal(days) / _YEAR).quantize(_CENTS, rounding=ROUND_HALF_UP)
        return InterestResponse(
            accountId=account_id, balance=balance, rate=rate, days=days, interest=value
        )
